types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+
+    for (unsigned int i = 0 ; i < (NN_+1)*nm_ ; i++){
+        
+        sol->z[i] = z[i];
+        sol->v[i] = v[i];
+        sol->lambda[i] = lambda[i];
+    
+    }
+
+    #endif
+
+    // Measure time
+    #if MEASURE_TIME == 1
+    
+    #if WIN32
+    QueryPerformanceCounter(&post_polish); // Get time after polishing
+    t_run_time = 1000000000ULL * (post_polish.QuadPart - start.QuadPart) / frequency.QuadPart;
+    t_polish_time = 1000000000ULL * (post_polish.QuadPart - post_solve.QuadPart) / frequency.QuadPart;
+    sol->run_time = t_run_time/(double)1e+9;
+    sol->polish_time = t_polish_time/(double)1e+9;
+    #else // If Linux
+    clock_gettime(CLOCK_MONOTONIC_RAW, &post_polish);
+    sol->run_time = (double) ( (post_polish.tv_sec - start.tv_sec) * 1000.0 ) + (double) ( (post_polish.tv_nsec - start.tv_nsec) / 1000000.0 );
+    sol->polish_time = (double) ( (post_polish.tv_sec - post_solve.tv_sec) * 1000.0 ) + (double) ( (post_polish.tv_nsec - post_solve.tv_nsec) / 1000000.0 );
+    #endif
+    
+    #endif
+
+}
+
+
+
+void solve_banded_Chol(const double (*Alpha)[nn_][nn_], const double (*Beta)[nn_][nn_], double *z, double *d){
+    
+    double sum = 0.0;
+
+    double y[(NN_+2)*nn_] = {0.0};
+    
+    // Forward substitution
+
+    for (unsigned int k=0 ; k < NN_+2 ; k++){
+        
+        for (unsigned int i = 0 ; i <= nn_-1 ; i++){
+
+            sum = 0.0;
+
+            for(unsigned int p = 0 ; p+1 <= i ; p++){
+
+                sum += Beta[k][p][i] * y[(k)*nn_+p]; 
+            
+            }
+
+            if (k>0){
+                    
+                for(unsigned int p = 0; p < nn_ ; p++){
+  
+                    sum += Alpha[k-1][p][i] * y[(k-1)*nn_+p];
+
+                }
+            
+            }
+            
+            y[(k)*nn_+i] = (d[(k)*nn_+i] - sum) * Beta[k][i][i]; // This is a division by the diagonal of Beta, but the diagonal of Beta is inverted, so we multiplity instead by the diagonal inverted
+
+        }
+
+    }
+
+    // Backward substitution
+
+    for (unsigned int k = NN_+2 ; k > 0 ; k--){
+
+        for(unsigned int i = nn_ ; i > 0 ; i--){
+
+            sum = 0.0;
+
+            for(unsigned int p = i+1 ; p <= nn_ ; p++){
+
+                sum += Beta[k-1][i-1][p-1] * z[(k-1)*nn_+(p-1)];
+
+            }
+            
+            if (k < NN_+2){
+                
+                for(unsigned int p = 0 ; p < nn_ ; p++){
+                
+                    sum += Alpha[k-1][i-1][p] * z[(k)*nn_+p];
+                
+                }
+            
+            }
+
+            z[(k-1)*nn_+i-1] = (y[(k-1)*nn_+i-1] - sum) * Beta[k-1][i-1][i-1]; // This is a division by the diagonal of Beta, but the diagonal of Beta is inverted, so we multiplity instead by the diagonal inverted
+
+        }
+
+    }
+    
+
+}
+
+void solve_banded_diag_sys(const double (*Q_rho_i)[nn_], const double (*R_rho_i)[mm_], const double (*S_rho_i)[mm_], const double (*T_rho_i)[nn_], double *z, double *d){
```

Comment:
The name of this function should probably be more specific, since it is not really solving a **generic** banded-diagonal system, but instead one which is formed by matrices `Q`, `R` in the diagonal and then by `T` and `S` in the tail end.

Reply:
Now it is changed to 'solve_banded_QRST_sys()'.